graph.h

```diff
@@ -14,6 +14,8 @@ class AbstractFactor {
   virtual double eval(const values &/*x*/);
   virtual values gradient_at(const values &/*x*/);
   virtual hessian hessian_at(const values &/*x*/);
+  // Used for trimming so the graph doesn't grow without bound
+  virtual bool shiftIndices(int poseSize, int firstPoseIdx);
```

Comment:
If these are supposed to be pure virtual functions with no implementation (I think in the `.cpp` file you make them all do false assertions so they can't be called) you should just put `= 0;` at the end of the line and omit the implementation entirely. Then anything that attempts to call the virtual functions won't even compile, rather than throwing a runtime error.
```suggestion
  virtual double eval(const values &/*x*/) = 0;
  virtual values gradient_at(const values &/*x*/) = 0;
  virtual hessian hessian_at(const values &/*x*/) = 0;
  // Used for trimming so the graph doesn't grow without bound
  virtual bool shiftIndices(int poseSize, int firstPoseIdx) = 0;
```